src/trainer.py

```diff
@@ -1094,35 +1095,105 @@ def _save_checkpoint(self, training_state, phase="pretrain", new_best_macro=Fals
             task_states["global"]["scheduler"] = sched_params
         else:
             task_states["global"]["scheduler"] = None
-        torch.save(
-            task_states,
-            os.path.join(
-                self._serialization_dir,
-                "task_state_{}_epoch_{}{}.th".format(phase, epoch, best_str),
-            ),
-        )
 
         metric_states = {}
         for metric_name, metric_info in self._metric_infos.items():
             metric_states[metric_name] = {}
             metric_states[metric_name]["hist"] = metric_info["hist"]
             metric_states[metric_name]["stopped"] = metric_info["stopped"]
             metric_states[metric_name]["best"] = metric_info["best"]
-        torch.save(
-            metric_states,
-            os.path.join(
-                self._serialization_dir,
-                "metric_state_{}_epoch_{}{}.th".format(phase, epoch, best_str),
-            ),
-        )
-        log.info("Saved to %s", self._serialization_dir)
+
+        if phase == "pretrain":
```

Comment:
It'd be way more readable if we split this into two methods, one (new) method for pretraining/mtl, and one (that we have now) for target tasks.

Reply:
Since they share some overlap, splitting the parts after this into two functions.